rfcs/0000-terraform-workflow-improvements.md

```diff
@@ -0,0 +1,826 @@
+- Module testing workspaces (CLI-driven), where application modules can be tested in isolation
+
+The base root modules (e.g. the current `prod` and `global`) will live in the Infrastructure team's environment, where all engineers will have some level of access to view runs and propose changes.
+
+### Proposed Workflow
+
+To carry out an infrastructure configuration change under the various workflow changes proposed in this document, an engineer will:
+
+1. Make changes to an application or infrastructure component module
+2. [Test those changes in isolation][reference-module-testing]
+3. Deploy those changes to their team's development/staging environment to verify their impact
+4. Deploy those changes to production once they are tested and approved
+
+Scalr will be an integral part of each step in this workflow, as detailed below.
+
+#### Scalr Workflow for Module Testing
+[scalr-workflow-module-testing]: #scalr-workflow-for-module-testing
+
+Each testing module will have a CLI-driven Scalr workspace with a remote operation backend that runs all `terraform` commands via Scalr. `terraform` commands will be run manually by the engineer, or possibly as part of a GitHub Actions workflow. Scalr's remote operation backend means the actual work will be done remotely by Scalr, using [Scalr's AWS permissions][reference-permissions], and the `terraform` CLI will show the run output locally.
+
+##### Continuous Integration
+
+- Engineer creates or modifies module configuration locally
+- Engineer pushes changes to GitHub branch
+- GitHub Actions workflow runs:
+  - `terraform validate`
+  - `terraform fmt`
+- Engineer (or GitHub Actions) tests module creation/destruction via Scalr (see workflow description in [Module Testing Workflow section][reference-module-testing])
+- Engineer opens pull request
+
+##### Code Review/Approval
+
+- Review &amp; approval handled via GitHub PR
+- Branch protection rules ensure that PR cannot be merged without approval
+
+##### Applying Changes
+
+- Engineer merges approved PR
+- If module versioning is employed, GitHub Actions triggers workflow to publish new version of module to module registry
+
+Once a new version of the module is published, the module version can be updated in the development/staging and production contexts using the workflows described below.
+
+If module versioning is not employed, changes to the module will need to be applied immediately in any dependent root modules. In the development/staging context, this can be handled automatically by configuring [run triggers](https://docs.scalr.com/en/latest/workspaces.html#run-triggers) in the Scalr workspace. In the production context, however, there will be an extra layer of protection to ensure that changes don't impact production before they're ready to be applied there. This is described in more detail in the [production workflow section][scalr-workflow-production] below.
```

Comment:
When would one module depend on another application module, and not be in the same terraform root module?

Reply:
I'm not sure I understand this question, but maybe it helps to clarify that the module testing workflow will only apply to _child_ modules, such as application modules and infrastructure component modules. Let me know if I should make that clearer in the text.